### PR TITLE
Reallocates the memory of all in-use arrays to mitigate fragmentation

### PR DESCRIPTION
**Pull request by sonots**
follow-up: https://github.com/cupy/cupy/pull/277

### CODE REVIEW

**Comment by sonots** [comment]
@anaruse can I ask you to review and check de-fragmentation by any chances?

**Comment by sonots** [comment]
Based on my experiments on train_imagenet.py, I could not get good improvement:

Before

```
out of memory to allocate 120422400 bytes (total 12602020864 bytes)
```

After

```
out of memory to allocate 120422400 bytes (total 12602020864 bytes)
```

So, no improvement. Hmm 🤔

**Comment by sonots** [comment]
Ah, my environemt was 12GB memory (I was thinking as 16G). Forget it.

File: cupy/cuda/memory.pyx [hunk 1]
```diff
@@ -702,6 +712,88 @@ cdef class SingleDeviceMemoryPool:
...
+                if new_memptr is None:
+                    return  # give up
+                if new_memptr.ptr > ptr:
+                    # This is based on an assumption that logical address are
+                    # almost linearly mapped to physical address. If the
+                    # assumption is correct, you can make larger continuous
+                    # free space in the high address area by calling the
+                    # realloc() twice when new address > old address. Note that
+                    # there is no guarantee that the assumption is correct.
+                    new_memptr_2nd = self._realloc(ptr, size)
```
Comment (anaruse) [inline cupy/cuda/memory.pyx]: This 2nd re-allocation should be tried after new chunk(s) is creased based on 1st re-allocation and original pointer is released. Or, you will not get expected results.

Comment (sonots) [reply cupy/cuda/memory.pyx]: okay, i will fix

**Comment by anaruse** [comment]
I checked how the memory usage change depending on whether re-allocation is used or not, using googlenet and GPU of 16GB memory. But no significant difference was observed, with or without realloc(). In either case, batch size can be increased up-to near 400.

First, I think think is great achievement of new memory pool since max batch size was about 300 when old memory pool was used. However, the batch size of 400 is slightly smaller than the max batch size when re-allocation approach was used with old memory pool.

So, I added some instrument codes to check the memory usage status and found that there are sort of memory fragmentation inside each `Memory` instance. For example, when trying to allocate 1GB of memory, although there was 2.3 GB free memory in total, they could not be released because they are distributed as `Chunk` instances over number of `Memory` instances, resulting in out-of-memory error.

Maybe, we need more sophisticated (complicated?) re-allocation method than can reduce its size by removing free chunks in each memory instance, in order to mitigate this fragmentation problem.

**Comment by sonots** [comment]
Thank you for your investigation!

> no significant difference was observed

This is same result with my experiment. I am relaxed since you also got same result.

> trying to allocate 1GB of memory, although there was 2.3 GB free memory in memory pool in total, they could not be released because they are distributed as Chunk instances over number of Memory instances, resulting in out-of-memory error.

I have ever thought that we may need memory **compaction** someday to mitigate fragmentation on **our** memory pool.

The logic will be different with this re-allocation patch.

**Comment by sonots** [comment]
I created another issue for memory compaction https://github.com/cupy/cupy/issues/536.
@anaruse do you agree we close this re-allocation patch?

**Comment by anaruse** [comment]
@sonots, Yes I agree with you.

**Comment by sonots** [comment]
thanks!

**Pull request closed without merging**